avm/res/azure-stack-hci/marketplace-gallery-image/main.bicep

```
@@ -0,0 +1,173 @@
+        }
+      }
+    }
+  }
+}
+
+resource marketplaceGalleryImages 'Microsoft.AzureStackHCI/marketplaceGalleryImages@2024-10-01-preview' = {
+  extendedLocation: {
+    name: customLocationResourceId
+    type: 'CustomLocation'
+  }
+  location: location
+  name: name
+  properties: marketplaceGalleryImageProperties
+  tags: tags
+}
+
+resource marketplaceGalleryImages_roleAssignments 'Microsoft.Authorization/roleAssignments@2022-04-01' = [
+  for (roleAssignment, index) in (formattedRoleAssignments ?? []): {
+    name: roleAssignment.?name ?? guid(
+      marketplaceGalleryImages.id,
+      roleAssignment.principalId,
+      roleAssignment.roleDefinitionId
+    )
+    properties: {
+      roleDefinitionId: roleAssignment.roleDefinitionId
+      principalId: roleAssignment.principalId
+      description: roleAssignment.?description
+      principalType: roleAssignment.?principalType
+      condition: roleAssignment.?condition
+      conditionVersion: !empty(roleAssignment.?condition) ? (roleAssignment.?conditionVersion ?? '2.0') : null // Must only be set if condtion is set
+      delegatedManagedIdentityResourceId: roleAssignment.?delegatedManagedIdentityResourceId
+    }
+    scope: marketplaceGalleryImages
+  }
+]
+
+// ============ //
+// Outputs      //
+// ============ //
+
+@description('The name of the marketplace gallery images.')
+output name string = marketplaceGalleryImages.name
+
+@description('The resource ID of the marketplace gallery images.')
+output resourceId string = marketplaceGalleryImages.id
+
+@description('The resource group of the marketplace gallery images.')
+output resourceGroupName string = resourceGroup().name
+
+@description('The location of the marketplace gallery images.')
+output location string = marketplaceGalleryImages.location
+
+// =============== //
+//   Definitions   //
+// =============== //
+
+type galleryOSDiskImageType = {
+  // Note: The API definition shows this as empty, so I'm defining it as an empty object
+}
+
+type galleryImageVersionStorageProfileType = {
+  @description('Required. This is the OS disk image.')
+  osDiskImage: galleryOSDiskImageType
+}
+
+type galleryImageVersionPropertiesType = {
+  @description('Required. This is the storage profile of a Gallery Image Version.')
+  storageProfile: galleryImageVersionStorageProfileType
+}
+
+type galleryImageVersionType = {
+  @description('Required. This is the version of the gallery image.')
+  name: string
+
+  @description('Optional. Describes the properties of a gallery image version.')
+  properties: galleryImageVersionPropertiesType?
+}
+
+type galleryImageIdentifierType = {
+  @description('Required. The name of the gallery image definition publisher.')
+  publisher: string
+
+  @description('Required. The name of the gallery image definition offer.')
+  offer: string
+
+  @description('Required. The name of the gallery image definition SKU.')
+  sku: string
+}
```

```suggestion
type galleryImageVersionType = {
  @description('Required. This is the version of the gallery image.')
  name: string

  @description('Optional. Describes the properties of a gallery image version.')
  properties: {
    @description('Required. This is the storage profile of a Gallery Image Version.')
    storageProfile: {
      @description('Required. This is the OS disk image.')
      osDiskImage: {
        @description('Optional. The size of the VHD to be created.')
        sizeInMB: int?
      }
    }
  }?
}

type galleryImageIdentifierType = {
  @description('Required. The name of the gallery image definition publisher.')
  publisher: string

  @description('Required. The name of the gallery image definition offer.')
  offer: string

  @description('Required. The name of the gallery image definition SKU.')
  sku: string
}
```

If `osDiskImage` is an empty type, it might have made sense to not surface it in general. But according to [this source](https://learn.microsoft.com/en-us/rest/api/stackhci/marketplace-gallery-images/create-or-update?view=rest-stackhci-2024-01-01&tabs=HTTP#galleryosdiskimage), there's actually something here: